apps/studio/hooks/ui/useConfirmOnClose.tsx

```diff
@@ -0,0 +1,69 @@
+import { useCallback, useState, type ReactNode } from 'react'
+
+export type ConfirmOnCloseModalProps = {
+  visible: boolean
+  onClose: () => void
+  onCancel: () => void
+}
+
+type UseConfirmOnCloseBaseParams = {
+  checkIsDirty: () => boolean
+  onClose: () => void
+}
+
+type UseConfirmOnCloseParams<ExtraProps extends Record<string, unknown> | undefined = undefined> =
+  UseConfirmOnCloseBaseParams &
+    (ExtraProps extends undefined
+      ? {
+          ConfirmationModal: (props: ConfirmOnCloseModalProps) => ReactNode
+        }
+      : {
+          ConfirmationModal: (props: ConfirmOnCloseModalProps & ExtraProps) => ReactNode
+          extraProps: ExtraProps
+        })
+
+type ConfirmOnCloseReturn = {
+  confirmOnClose: () => void
+  modal: ReactNode
+}
+
+export const useConfirmOnClose = <
+  ExtraProps extends Record<string, unknown> | undefined = undefined,
+>(
+  props: UseConfirmOnCloseParams<ExtraProps>
+): ConfirmOnCloseReturn => {
+  const { checkIsDirty, onClose, ConfirmationModal } = props
+
+  const [visible, setVisible] = useState(false)
+
+  const confirmOnClose = useCallback(() => {
+    if (checkIsDirty()) {
+      setVisible(true)
+    } else {
+      onClose()
+    }
+  }, [checkIsDirty, onClose])
```

Comment:
`useCallback()` is probably negatively impacting performance here since `checkIsDirty` and `onClose` are almost never memoized. We can still use it though if you wrap them both with `useLatest()` and remove them from the dependency array.